### ISSUE TITLE
[java] UseEqualsToCompareStrings: FP with "var" variables

### ISSUE THREAD

**Issue by kris-scheibe** [pmd/pmd#2979]
**Affects PMD Version:** 6.30.0

**Rule:** [UseEqualsToCompareStrings](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#useequalstocomparestrings)

**Description:**
The rule detects a violation even though not two String variables but the results of the ".isEmpty()" method calls are compared.

This might be related to #2976 as CompareObjectsWithEquals falsely detects this as well.

Neither of the violations are reported if the variables are declared as "String" instead of "var".

**Code Sample demonstrating the issue:**

```java
final Matcher matcher = ...;

if (matcher.matches()) {

final var firstString = matcher.group("a");
final var secondString = matcher.group("b");

if (firstString.isEmpty() != secondString.isEmpty()) { // <- violation
    // ...
}

}
```

**Expected outcome:**

*   The compared values are boolean, not Strings -> false-positive

**Running PMD through:** *Other*

**Comment by oowekyala** [comment]
The type of those variables is inferred to Matcher for some reason...

**Comment by catellie** [comment]
I think I've just seen a variant of this when I tried to upgrade to 6.30:

Suddenly I see false positives for code like this:

`
String a = "";
String b = "";
if (a.length() == b.length() { // <-- this line flags UseEqualsToCompareStrings
}

`
